MoreLinq/UnFold.cs

```diff
@@ -0,0 +1,111 @@
+        /// Function to select the next state.
+        /// </param>
+        /// <param name="resultSelector">
+        /// Function to select the next result.
+        /// </param>
+        /// <returns>A sequence containing the results generated by the resultSelector function.</returns>
+        /// <remarks>
+        /// This operator uses deferred execution and streams its results.
+        /// </remarks>
+
+        public static IEnumerable<TResult> Unfold<TState, T, TResult>(
+            TState state,
+            Func<TState, T> generator,
+            Func<T, bool> predicate,
+            Func<T, TState> stateSelector,
+            Func<T, TResult> resultSelector)
+        {
+            if (generator == null) throw new ArgumentNullException(nameof(generator));
+            if (predicate == null) throw new ArgumentNullException(nameof(predicate));
+            if (stateSelector == null) throw new ArgumentNullException(nameof(stateSelector));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            return UnfoldImpl(state, generator, predicate, stateSelector, resultSelector);
+        }
+
+        /// <summary>
+        /// Returns a sequence of values generated by a generator function, which receives a state
+        /// and returns either null if it is done producing the sequence, or a tuple with the next element
+        /// of the sequence and the value used as the next state in the recursive call.
+        /// </summary>
+        /// <typeparam name="TState">Type of state elements.</typeparam>
+        /// <typeparam name="T">The type of the elements of the result sequence.</typeparam>
+        /// <param name="state">The initial state value.</param>
+        /// <param name="generator">
+        /// Function that takes a state and returns a tuple of the next element of the sequence
+        /// and the next state. If the function returns null, it is done producing the sequence.
+        /// </param>
+        /// <returns>A sequence containing the results generated by the generator function.</returns>
+        /// <remarks>
+        /// This operator uses deferred execution and streams its results.
+        /// </remarks>
+
+        public static IEnumerable<T> Unfold<TState, T>(TState state, Func<TState, Tuple<T, TState>> generator)
+        {
+            if (generator == null) throw new ArgumentNullException(nameof(generator));
+
+            return UnfoldImpl(state, generator, e => e != null, e => e.Item2, e => e.Item1);
+        }
+
+
+        private static IEnumerable<TResult> UnfoldImpl<TState, T, TResult>(
+            TState state,
+            Func<TState, T> generator,
+            Func<T, bool> predicate,
+            Func<T, TState> stateSelector,
+            Func<T, TResult> resultSelector)
+        {
+            while (true)
+            {
+                var step = generator(state);
+
+                if (!predicate(step))
```

Comment:
> BTW, we should have a test for this behavior too.

Agreed! I haven't been over the tests or any of the PR files since I'm still waiting for the [build to be fixed](https://github.com/morelinq/MoreLINQ/pull/260#pullrequestreview-25443733) first. 😆